fem/adnonlininteg.hpp

```diff
@@ -0,0 +1,384 @@
+      {
+         ADFVector aduu(uu); //all dual numbers are initialized to zero
+         ADFType rez;
+
+         for (int ii = 0; ii < n; ii++)
+         {
+            aduu[ii].dual(1.0);
+            rez = func(vparam, aduu);
+            rr[ii] = rez.dual();
+            aduu[ii].dual(0.0);
+         }
+      }
+#endif
+   }
+
+   void QFunctionDU(const Vector &vparam, Vector &uu, Vector &rr)
+   {
+      CTD<double, Vector> func;
+      func(vparam, uu, rr);
+   }
+
+   void QFunctionDD(const Vector &vparam, Vector &uu, DenseMatrix &jac)
+   {
+#if defined MFEM_USE_ADEPT
+      //use ADEPT package
+      adept::Stack *p_stack = adept::active_stack();
+      p_stack->deactivate();
+
+      int n = uu.Size();
+      jac.SetSize(m, n);
+      jac = 0.0;
+      m_stack.activate();
+      {
+         ADFVector aduu(uu);
+         ADFVector rr(m); //residual vector
+         m_stack.new_recording();
+         QFunctionDU(vparam, aduu, rr);
+         m_stack.independent(aduu.GetData(), n); //independent variables
+         m_stack.dependent(rr.GetData(), m);     //dependent variables
+         m_stack.jacobian(jac.Data());
+      }
+      m_stack.deactivate();
+#elif defined MFEM_USE_FADBADPP
+      //use FADBAD++
+#ifdef MFEM_USE_ADFORWARD
+      int n = uu.Size();
+      jac.SetSize(m, n);
+      jac = 0.0;
+      {
+         ADFVector aduu(uu);
+         ADFVector rr(m);
+
+         for (int ii = 0; ii < n; ii++)
+         {
+            aduu[ii].diff(ii, n);
+         }
+         QFunctionDU(vparam, aduu, rr);
+         for (int ii = 0; ii < n; ii++)
+         {
+            for (int jj = 0; jj < m; jj++)
+            {
+               jac(jj, ii) = rr[jj].d(ii);
+            }
+         }
+      }
+#else
+      int n = uu.Size();
+      jac.SetSize(m, n);
+      jac = 0.0;
+      {
+         ADFVector aduu(uu);
+         ADFVector rr(m);
+         QFunctionDU(vparam, aduu, rr);
+         for (int ii = 0; ii < m; ii++)
+         {
+            rr[ii].diff(ii, m);
+         }
+         for (int ii = 0; ii < n; ii++)
+         {
+            for (int jj = 0; jj < m; jj++)
+            {
+               jac(jj, ii) = aduu[ii].d(jj);
+            }
+         }
+      }
+#endif
+#else
+      //use native AD package
+      int n = uu.Size();
+      jac.SetSize(m, n);
+      jac = 0.0;
+      {
+         ADFVector aduu(uu); //all dual numbers are initialized to zero
+         ADFVector rr(m);
+
+         for (int ii = 0; ii < n; ii++)
+         {
+            aduu[ii].dual(1.0);
+            QFunctionDU(vparam, aduu, rr);
+            for (int jj = 0; jj < m; jj++)
+            {
+               jac(jj, ii) = rr[jj].dual();
+            }
+            aduu[ii].dual(0.0);
+         }
+      }
+#endif
+   }
+};
+
+//template class for differentiation; the function
+//for differentiation is supplied as a functor
+//the operator()(scalar,vector) defines the actual function
+template<template<typename, typename> class CTD>
+class ADQFunctionTH
```

Comment:
This needs proper documentation.

Reply:
I think it also needs a readme with how you expect a user to actually use it.

Reply:
Or at least a large comment block.

Reply:
Documented! Please, return comments!